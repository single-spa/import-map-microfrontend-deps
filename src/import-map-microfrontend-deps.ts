import { Generator, Install } from "@jspm/generator";
import { writeFile, mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import { ImportMapMicrofrontendUtils } from "@single-spa/import-map-microfrontend-utils";

interface ImportMapTemplate {
  imports: ModuleMap;
  scopes: Record<string, ModuleMap>;
}

interface ModuleMap {
  [specifier: string]: string;
}

interface ImportMapBuildOptions {
  template: ImportMapTemplate;
  outputFolder: string;
  utils: ImportMapMicrofrontendUtils;
}

export async function buildImportMapDependencies(
  options: ImportMapBuildOptions,
) {
  const { template } = options;

  const generatorOptions = {
    integrity: true,
  };

  const importMapGenerator = new Generator(generatorOptions);

  // Install shared dependencies into import map
  console.log("Creating import map from template", template);
  for (const importMapKey in template.imports) {
    await importMapGenerator.install(
      processImportMapEntry(importMapKey, template.imports[importMapKey]),
    );
  }

  const scopeGenerators = {};

  for (const microfrontendName in template.scopes ?? {}) {
    const scopeGenerator = new Generator(generatorOptions);
    for (const importMapKey in template.scopes[microfrontendName]) {
      await scopeGenerator.install(
        processImportMapEntry(
          importMapKey,
          template.scopes[microfrontendName][importMapKey],
        ),
      );
    }

    scopeGenerators[microfrontendName] = scopeGenerator;
  }

  const finalMap = importMapGenerator.getMap();
  if (!finalMap.imports) {
    finalMap.imports = {};
  }

  if (!finalMap.scopes) {
    finalMap.scopes = {};
  }

  for (const microfrontendName in scopeGenerators) {
    const generator = scopeGenerators[microfrontendName];
    const scopeMap = generator.getMap();

    finalMap.scopes[
      options.utils.getMicrofrontendURLPrefix(microfrontendName)
    ] = {
      ...(finalMap.scopes[microfrontendName] ?? {}),
      // TODO - account for its scopes
      ...scopeMap.imports,
    };

    finalMap.integrity = {
      ...finalMap.integrity,
      ...scopeMap.integrity,
    };
  }

  // Replace public CDN urls with self-hosted urls in import map
  const importMapString = JSON.stringify(finalMap, null, 2).replace(
    /https:\/\/ga.jspm.io\//g,
    options.utils.getDependenciesURLPrefix(),
  );

  console.log(importMapString);

  // Write final import map
  await mkdir(options.outputFolder, { recursive: true });
  await writeFile(
    `${options.outputFolder}/deps.importmap`,
    importMapString,
    "utf-8",
  );

  const downloadedUrls = [];

  for (let url in finalMap.integrity) {
    if (downloadedUrls.includes(url)) {
      continue;
    }
    console.log(url);
    const r = await fetch(url);
    if (r.ok) {
      const filePath = `./${options.outputFolder}/${options.utils.getDependenciesFolderName()}/${url.replace("https://ga.jspm.io/", "")}`;
      const dir = dirname(filePath);
      await mkdir(dir, { recursive: true });
      await writeFile(filePath, await r.text(), "utf-8");
      downloadedUrls.push(url);
    } else {
      throw Error(`Unable to download file from JSPM CDN - url '${url}'`);
    }
  }
}

function processImportMapEntry(
  importMapKey: string,
  importMapValue: string,
): Install {
  const splitKey = importMapKey.split("/");
  const extraSlashes = splitKey.length - (importMapKey.startsWith("@") ? 2 : 1);

  const result: Install = {
    target: `${splitKey.slice(0, splitKey.length - extraSlashes).join("/")}@${importMapValue}`,
  };

  if (extraSlashes > 0) {
    result.subpath = `./${splitKey.slice(extraSlashes).join("/")}`;
  }

  return result;
}
