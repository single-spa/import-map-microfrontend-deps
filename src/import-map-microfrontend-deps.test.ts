import { beforeEach, test, vi, expect } from "vitest";
import { buildImportMapDependencies } from "./import-map-microfrontend-deps";
import { ImportMapMicrofrontendUtils } from "@single-spa/import-map-microfrontend-utils";
import { vol, fs } from "memfs";
import path from "path";

vi.mock("node:fs/promises");

beforeEach(() => {
  vol.reset();
});

test(`writes an import map and dependencies to the file system`, async () => {
  await buildImportMapDependencies({
    template: {
      imports: {
        react: "18.3.0",
        "react-dom": "18.3.0",
      },
      scopes: {},
    },
    outputFolder: "dist",
    utils: new ImportMapMicrofrontendUtils({
      baseOrigin: "https://cdn.example.com",
    }),
  });

  expect(
    await fs.readFileSync("./dist/deps.importmap", "utf-8"),
  ).toMatchSnapshot();

  expect(
    await fs.readdirSync("./dist/deps", { recursive: true }),
  ).toMatchSnapshot();
});

test(`scoped microfrontend dependencies`, async () => {
  await buildImportMapDependencies({
    template: {
      imports: {
        react: "18.3.0",
        "react-dom": "18.3.0",
      },
      scopes: {
        "@org/navbar": {
          react: "17.0.0",
          "react-dom": "17.0.0",
        },
      },
    },
    outputFolder: "dist",
    utils: new ImportMapMicrofrontendUtils({
      baseOrigin: "https://cdn.example.com",
    }),
  });

  expect(
    await fs.readFileSync("./dist/deps.importmap", "utf-8"),
  ).toMatchSnapshot();

  const files = (await fs.readdirSync("./dist/deps", { recursive: true })).map(
    (file) => path.relative(process.cwd(), file),
  );
  expect(files).toMatchSnapshot();
});
