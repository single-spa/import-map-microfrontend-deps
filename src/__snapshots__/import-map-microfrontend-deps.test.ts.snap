// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scoped microfrontend dependencies 1`] = `
"{
  "imports": {
    "react": "https://cdn.example.com/deps/npm:react@18.3.0/dev.index.js",
    "react-dom": "https://cdn.example.com/deps/npm:react-dom@18.3.0/dev.index.js"
  },
  "scopes": {
    "https://cdn.example.com/deps/": {
      "scheduler": "https://cdn.example.com/deps/npm:scheduler@0.23.2/dev.index.js"
    },
    "https://cdn.example.com/mfes/org-navbar/": {
      "react": "https://cdn.example.com/deps/npm:react@17.0.0/dev.index.js",
      "react-dom": "https://cdn.example.com/deps/npm:react-dom@17.0.0/dev.index.js"
    }
  },
  "integrity": {
    "https://cdn.example.com/deps/npm:react-dom@18.3.0/_/hZ0Hd0FS.js": "sha384-wtEbXP4B7Gl5fUVIN7xE2UNOcdbjiNLtkyBx4QFpyk6/HQXz0yErwNlwpi+JwKio",
    "https://cdn.example.com/deps/npm:react-dom@18.3.0/dev.index.js": "sha384-RlA1cs4/0fqgNMuo0f8Af1dWFwZUSBaF6F+UgrlQFXjv0em+AZI16QhG96aNaygO",
    "https://cdn.example.com/deps/npm:react@18.3.0/dev.index.js": "sha384-FqpzZfi2YeXImQD1xnS/wUNXnRBozQ7uHQmvA6RXbLiApqeXvNFSGQ8coq4Twfmd",
    "https://cdn.example.com/deps/npm:scheduler@0.23.2/dev.index.js": "sha384-RmHcVPSidbQWYik0CCkT1HaH1CbRQB1Z+/gt66Og9IhFxSqa4PNvCBUneAQl0I1R",
    "https://cdn.example.com/deps/npm:object-assign@4.1.1/index.js": "sha384-iQp1zoaqIhfUYyYkz3UNk1QeFfmBGgt1Ojq0kZD5Prql1g7fgJVzVgsjDoR65lv8",
    "https://cdn.example.com/deps/npm:react-dom@17.0.0/cjs/dev.react-dom.development.js": "sha384-jIY+OjbqIxEiVs6PBTAVxNsOEsryYSoAtv8azlFjCBTKRzG2QXQtAHAPeffEYjlR",
    "https://cdn.example.com/deps/npm:react-dom@17.0.0/dev.index.js": "sha384-aXY4KO6iA0Zb41YTFkV2cVsKafk4DiiksnE97768LXtZidU/HnB5LYL9725v9Ksw",
    "https://cdn.example.com/deps/npm:react@17.0.0/cjs/dev.react.development.js": "sha384-HStEslgnosSd15d5hs9LNsbUb2gDNL1dVDl8xVpIjM3UCc8WjUphSg3oyGLI8sjm",
    "https://cdn.example.com/deps/npm:react@17.0.0/dev.index.js": "sha384-X8NBRDFZMbZO1Q5UEe+S+jU2F/TGWPNWBcTdzeEhfbbuk9rBzU1+/c6tsqxvE3BG",
    "https://cdn.example.com/deps/npm:scheduler@0.20.2/_/b59585b3.js": "sha384-aPA2DgvtbHjaEfNi8sO1P0f4Dk2mfuYVDQm1TB0PbVph5kwqgq8V7wsL5G6nqVol",
    "https://cdn.example.com/deps/npm:scheduler@0.20.2/dev.index.js": "sha384-qF0Jy83btjdPADN4QLKKmk/aUUyJnDqT+kYomKiUQk4nWrBsHVkM67Pua+8nHYUt",
    "https://cdn.example.com/deps/npm:scheduler@0.20.2/dev.tracing.js": "sha384-2F0quaGupcgfUEwIGZsencGJp4bSamvXDI8mjsLg9ZXJeDDmGAL6US1H/ZLFGDpP"
  }
}"
`;

exports[`scoped microfrontend dependencies 2`] = `Promise {}`;

exports[`writes an import map and dependencies to the file system 1`] = `
"{
  "imports": {
    "react": "https://cdn.example.com/deps/npm:react@18.3.0/dev.index.js",
    "react-dom": "https://cdn.example.com/deps/npm:react-dom@18.3.0/dev.index.js"
  },
  "scopes": {
    "https://cdn.example.com/deps/": {
      "scheduler": "https://cdn.example.com/deps/npm:scheduler@0.23.2/dev.index.js"
    }
  },
  "integrity": {
    "https://cdn.example.com/deps/npm:react-dom@18.3.0/_/hZ0Hd0FS.js": "sha384-wtEbXP4B7Gl5fUVIN7xE2UNOcdbjiNLtkyBx4QFpyk6/HQXz0yErwNlwpi+JwKio",
    "https://cdn.example.com/deps/npm:react-dom@18.3.0/dev.index.js": "sha384-RlA1cs4/0fqgNMuo0f8Af1dWFwZUSBaF6F+UgrlQFXjv0em+AZI16QhG96aNaygO",
    "https://cdn.example.com/deps/npm:react@18.3.0/dev.index.js": "sha384-FqpzZfi2YeXImQD1xnS/wUNXnRBozQ7uHQmvA6RXbLiApqeXvNFSGQ8coq4Twfmd",
    "https://cdn.example.com/deps/npm:scheduler@0.23.2/dev.index.js": "sha384-RmHcVPSidbQWYik0CCkT1HaH1CbRQB1Z+/gt66Og9IhFxSqa4PNvCBUneAQl0I1R"
  }
}"
`;

exports[`writes an import map and dependencies to the file system 2`] = `Promise {}`;
